File: Question_2/routingSystem.kt

    data class Location(val lat: Double, val lon: Double)
    data class Student(val id: Int, val location: Location)
    data class BusStop(val id: String, val location: Location)

    fun euclidean(p1: Location, p2: Location): Double {
        val dx = p1.lat - p2.lat
        val dy = p1.lon - p2.lon
        return Math.sqrt(dx * dx + dy * dy)
    }

    fun main() {
        val students = listOf(
            Student(1, Location(1.290270, 36.821946)),
            Student(2, Location(1.292030, 36.822000)),
            Student(3, Location(1.289000, 36.819000)),
        )
        val busStops = listOf(
            BusStop("A", Location(1.291000, 36.820000)),
            BusStop("B", Location(1.293000, 36.823000)),
        )
        val busCapacity = 20
        val studentBus = mutableMapOf<Int,String>()
        for (student in students) {
            val nearest = busStops.minByOrNull { euclidean(student.location, it.location) }!!
            studentBus.put(student.id, nearest.id)
        }

        val buses = mutableMapOf<String, MutableList<Int>>()
        for (student in students) {
            val nearest = busStops.minByOrNull { euclidean(student.location, it.location) }!!
            buses.getOrPut(nearest.id) { mutableListOf() }.add(student.id)
        }

        for ((stop, group) in buses) {
            val grouped = group.chunked(busCapacity)
            println("Stop $stop buses: $grouped")
        }

        println(buses)
        println(studentBus)
    }

File: Question_2/routing_system.py
students = [
    {"id": 1, "location": (1.290270, 36.821946)},
    {"id": 2, "location": (1.292030, 36.822000)},
    {"id": 3, "location": (1.289000, 36.819000)},
]
bus_stops = [
    {"id": "A", "location": (1.291000, 36.820000)},
    {"id": "B", "location": (1.293000, 36.823000)},
]
bus_capacity = 35
buses = {}
def euclidean(p1, p2):
    return ((p1[0] - p2[0])**2 + (p1[1] - p2[1])**2) ** 0.5
for student in students:
    nearest = min(bus_stops, key=lambda stop: euclidean(student["location"], stop["location"]))
    stop_id = nearest["id"]
    if stop_id not in buses:
        buses[stop_id] = []
    buses[stop_id].append(student["id"])
for stop, group in buses.items():
    grouped = [group[i:i + bus_capacity] for i in range(0, len(group), bus_capacity)]
    print(f"Stop {stop} buses: {grouped}")
